drivers/gpu/drm/panel/panel-novatek-nt36672a.c

```diff
@@ -79,23 +71,740 @@ static inline struct nt36672a_panel *to_nt36672a_panel(struct drm_panel *panel)
+
+	/* Send rest of the init cmds (.on_cmds_2) */
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF, 0x24);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFB, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC3, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC4, 0x54);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF, 0x10);
+
+	mipi_dsi_msleep(&dsi_ctx, 120);
+
+	return dsi_ctx.accum_err;
+}
+
+/* Tianma-beryllium specific handler with specific delays */
+static int tianma_beryllium_off_cmds(struct drm_panel *panel)
+{
+	struct nt36672a_panel *pinfo = to_nt36672a_panel(panel);
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = pinfo->link };
 
-		if (err < 0)
-			return err;
+	/* send off cmds */
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF, 0x24);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFB, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xC3, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xFF, 0x10);
+
+	mipi_dsi_dcs_set_display_off_multi(&dsi_ctx);
+
+	/* 120ms delay required here as per DCS spec */
+	mipi_dsi_msleep(&dsi_ctx, 120);
+
+	mipi_dsi_dcs_enter_sleep_mode_multi(&dsi_ctx);
+
+	/* 0x3C = 60ms delay */
+	mipi_dsi_msleep(&dsi_ctx, 60);
+
+	return dsi_ctx.accum_err;
+}
+
+static int tianma_lavender_init_cmds(struct drm_panel *panel)
+{
+	struct nt36672a_panel *pinfo = to_nt36672a_panel(panel);
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = pinfo->link };
+
+	pinfo->link->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xff, 0x25);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfb, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x18, 0x96);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x05, 0x04);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xff, 0x20);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfb, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x78, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xff, 0x24);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfb, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x82, 0x13);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x84, 0x31);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x88, 0x13);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x8a, 0x31);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x8e, 0xe4);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x8f, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x90, 0x80);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xff, 0x26);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfb, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, MIPI_DCS_READ_PPS_CONTINUE, 0x12);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xaa, 0x10);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xae, 0x8a);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xff, 0x10);
+
+	mipi_dsi_dcs_exit_sleep_mode_multi(&dsi_ctx);
+	mipi_dsi_msleep(&dsi_ctx, 80);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xb0, 0x01);
+	mipi_dsi_dcs_set_tear_on_multi(&dsi_ctx, MIPI_DSI_DCS_TEAR_MODE_VBLANK);
+
+	if (of_device_is_compatible(dev_of_node(&pinfo->link->dev),
+				    "novatek,nt36672a-tianma-lavender")) {
```

Comment:
The check for compatible string here is now incorrect, should be "tianma,tl063fvmca01-00" (lavender-tianma panel still worked though :laughing:  )

3 panels (lavender-tianma, lavender-shenchao, tulip) can and probaly should use the same init function, and be added in a one single commit in preparation for submission upstream.